lib/engine/game/g_1832/step/token.rb

```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+require_relative '../../../step/token'
+
+module Engine
+  module Game
+    module G1832
+      module Step
+        class Token < Engine::Step::Token
+          def actions(entity)
+            actions = super.dup
+            actions += %w[choose pass] if can_place_miami_token?(entity)
+
+            actions.uniq
+          end
+
+          def choice_name
+            'Additional Token Actions'
+          end
+
+          def choices
+            choices = []
+            choices << ['Place Key West Token'] if can_place_miami_token?(current_entity)
+            choices
+          end
+
+          def process_choose(action)
+            place_miami_token(action.entity) if action.choice == 'Place Key West Token'
+            pass!
+          end
+
+          def place_miami_token(corporation)
+            ability = @game.abilities(corporation, :assign_hexes)
+            hex = @game.hex_by_id(ability.hexes.first)
+
+            hex.tile.icons.reject! { |icon| icon.name == 'FECR_key_west' }
+            hex.assign!(corporation)
+            ability.use!
+            corporation.spend(ability.cost, @game.bank)
+            @log << "#{corporation.name} places the Key West token on #{hex.name} for #{@game.format_currency(ability.cost)}"
+          end
+
+          def skip!
+            pass!
+          end
+
+          def can_place_miami_token?(entity)
+            (ability = @game.abilities(entity, :assign_hexes)) &&
```

Comment:
If there any chance that `entity` could have more than one `:assign_hexes` ability? If it does, then this code is going to cause errors: `@game.abilities` can either return a single `Ability` or an array of them (or nil). Your code will break if it gets an array.

Reply:
I borrowed this code from elsewhere, hadn't even considered that. Updated to clearly find this ability.

>           def can_place_miami_token?(entity)
>             (abilities = @game.abilities(entity, :assign_hexes)) &&
>             abilities.any? { |ability| ability.cost == 100 } &&
>             entity.cash >= 100 &&
>             (3..6).cover?(@game.phase.name.to_i)
>           end

Reply:
Now you are assuming that you'll get an array returned from `@game.abilities`. If there's just one matching ability it returns an `Ability` object, not an array.
https://github.com/tobymao/18xx/blob/d5e0e80e7b870eccefd5bef31930818fccc705ff/lib/engine/game/base.rb#L1975-L1996
If you're not absolutely sure whether there will be none, one or more matching abilities returned from `@game.abilities` then convert the returned value to an array:
```
@abilities = Array(@game.abilities(entity, :assign_hexes)
```

Reply:
awesome, thanks!